File: test/warnings/sql_tables_warning/python/SqlTablesWarningTest.py
import unittest

from testutils import getZserioApi, assertWarningsPresent

class SqlTablesWarningTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.warnings = {}
        cls.api = getZserioApi(__file__, "sql_tables_warning.zs",
                               expectedWarnings=7, errorOutputDict=cls.warnings)

    def testBadOrderedPrimaryKey(self):
        assertWarningsPresent(self,
            "sql_tables_warning.zs",
            [
                "bad_ordered_primary_key_warning.zs:9:9: "
                "Primary key column 'classId' is in bad order in sql table 'BadOrderedPrimaryKeyTable'."
            ]
        )

    def testDuplicatedPrimaryKey(self):
        assertWarningsPresent(self,
            "sql_tables_warning.zs",
            [
                "duplicated_primary_key_warning.zs:6:33: "
                "Duplicated primary key column 'classId' in sql table 'DuplicatedPrimaryKeyTable'."
            ]
        )

    def testMultiplePrimaryKeys(self):
        assertWarningsPresent(self,
            "sql_tables_warning.zs",
            [
                "multiple_primary_keys_warning.zs:9:9: "
                "Multiple primary keys in sql table 'MultiplePrimaryKeysTable'."
            ]
        )

    def testNoPrimaryKey(self):
        assertWarningsPresent(self,
            "sql_tables_warning.zs",
            [
                "no_primary_key_warning.zs:3:11: "
                "No primary key in sql table 'NoPrimaryKeyTable'."
            ]
        )

    def testNotFirstPrimaryKey(self):
        assertWarningsPresent(self,
            "sql_tables_warning.zs",
            [
                "not_first_primary_key_warning.zs:6:29: "
                "Primary key column 'classId' is not the first one in sql table 'NotFirstPrimaryKeyTable'."
            ]
        )

    def testNotNullPrimaryKey(self):
        assertWarningsPresent(self,
            "sql_tables_warning.zs",
            [
                "not_null_primary_key_warning.zs:5:17: "
                "Primary key column 'schoolId' can contain NULL in sql table 'NotNullPrimaryKeyTable1'."
            ]
        )

        assertWarningsPresent(self,
            "sql_tables_warning.zs",
            [
                "not_null_primary_key_warning.zs:14:17: "
                "Primary key column 'schoolId' can contain NULL in sql table 'NotNullPrimaryKeyTable2'."
            ]
        )

File: test/warnings/sql_tables_warning/cpp/SqlTablesWarningTest.cpp
#include <string>
#include <vector>

#include "gtest/gtest.h"

#include "test_utils/ZserioErrorOutput.h"

class SqlTablesWarningTest : public ::testing::Test
{
protected:
    SqlTablesWarningTest()
    :   zserioWarnings("warnings/sql_tables_warning")
    {}

    const test_utils::ZserioErrorOutput zserioWarnings;
};

TEST_F(SqlTablesWarningTest, badOrderedPrimaryKey)
{
    const std::string warning = "bad_ordered_primary_key_warning.zs:9:9: "
            "Primary key column 'classId' is in bad order in sql table 'BadOrderedPrimaryKeyTable'.";
    ASSERT_TRUE(zserioWarnings.isPresent(warning));
}

TEST_F(SqlTablesWarningTest, duplicatedPrimaryKey)
{
    const std::string warning = "duplicated_primary_key_warning.zs:6:33: "
            "Duplicated primary key column 'classId' in sql table 'DuplicatedPrimaryKeyTable'.";
    ASSERT_TRUE(zserioWarnings.isPresent(warning));
}

TEST_F(SqlTablesWarningTest, multiplePrimaryKeys)
{
    const std::string warning = "multiple_primary_keys_warning.zs:9:9: "
            "Multiple primary keys in sql table 'MultiplePrimaryKeysTable'.";
    ASSERT_TRUE(zserioWarnings.isPresent(warning));
}

TEST_F(SqlTablesWarningTest, noPrimaryKey)
{
    const std::string warning = "no_primary_key_warning.zs:3:11: "
            "No primary key in sql table 'NoPrimaryKeyTable'.";
    ASSERT_TRUE(zserioWarnings.isPresent(warning));
}

TEST_F(SqlTablesWarningTest, notFirstPrimaryKey)
{
    const std::string warning = "not_first_primary_key_warning.zs:6:29: "
            "Primary key column 'classId' is not the first one in sql table 'NotFirstPrimaryKeyTable'.";
    ASSERT_TRUE(zserioWarnings.isPresent(warning));
}

TEST_F(SqlTablesWarningTest, notNullPrimaryKey)
{
    const std::string warning1 = "not_null_primary_key_warning.zs:5:17: "
            "Primary key column 'schoolId' can contain NULL in sql table 'NotNullPrimaryKeyTable1'.";
    ASSERT_TRUE(zserioWarnings.isPresent(warning1));

    const std::string warning2 = "not_null_primary_key_warning.zs:14:17: "
            "Primary key column 'schoolId' can contain NULL in sql table 'NotNullPrimaryKeyTable2'.";
    ASSERT_TRUE(zserioWarnings.isPresent(warning2));
}
